File: framework/flink/tour/java/write.java
// Write to CrateDB using Apache Flink.
// Invoke: uvx crash -c 'CREATE TABLE person (name STRING, age INT);'
// Invoke: gradle run write
// Source: https://github.com/crate/cratedb-examples/blob/main/framework/flink/tour/java/write.java

// https://nightlies.apache.org/flink/flink-docs-master/docs/connectors/datastream/jdbc/#full-example

import org.apache.flink.connector.jdbc.*;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;

public class write {

    public static void main(String[] args) throws Exception {
        var env = StreamExecutionEnvironment.getExecutionEnvironment();

        // Define source data.
        env.fromElements(
            new Person("Fred", 35),
            new Person("Wilma", 35),
            new Person("Pebbles", 2)

        // Define CrateDB as data sink.
        ).addSink(
            JdbcSink.sink(
                "INSERT INTO person (name, age) VALUES (?, ?)",
                (statement, person) -> {
                    statement.setString(1, person.name);
                    statement.setInt(2, person.age);
                },
                JdbcExecutionOptions.builder()
                    .withBatchSize(1000)
                    .withBatchIntervalMs(200)
                    .withMaxRetries(5)
                    .build(),
                new JdbcConnectionOptions.JdbcConnectionOptionsBuilder()
                    .withUrl("jdbc:postgresql://localhost:5432/doc?sslmode=disable")
                    .withDriverName("org.postgresql.Driver")
                    .withUsername("crate")
                    .withPassword("crate")
                    .build()
            ));

        // Execute pipeline.
        env.execute();
    }

    public static class Person {
        public String name;
        public Integer age;
        public Person() {}
        public Person(String name, Integer age) {
            this.name = name;
            this.age = age;
        }
    }

}

File: framework/flink/tour/python/write.py
import dataclasses
import logging

from pathlib import Path

from pyflink.common import Types
from pyflink.datastream import StreamExecutionEnvironment
from pyflink.datastream.connectors.jdbc import JdbcSink, JdbcConnectionOptions, JdbcExecutionOptions


logger = logging.getLogger(__name__)

JARS_PATH = Path(__file__).parent / 'jars'


@dataclasses.dataclass
class Person:
    name: str
    age: int


def main():

    env = StreamExecutionEnvironment.get_execution_environment()
    jars = list(map(lambda x: 'file://' + str(x), (JARS_PATH.glob('*.jar'))))
    env.add_jars(*jars)

    # Define source data.
    ds = env.from_collection([
        Person("Fred", 35),
        Person("Wilma", 35),
        Person("Pebbles", 2),
    ])

    # Define CrateDB as data sink.
    row_type_info = Types.ROW_NAMED(['name', 'age'], [Types.STRING(), Types.INT()])
    ds.add_sink(
        JdbcSink.sink(
            "INSERT INTO person (name, age) VALUES (?, ?)",
            row_type_info,

            # FIXME (Flink >= 1.19): java.lang.NoSuchMethodException: org.apache.flink.connector.jdbc.internal.JdbcOutputFormat.createRowJdbcStatementBuilder

            # This is due to a bug in the python Flink library. In `flink-connector-jdbc` v3.1,
            # the `JdbcOutputFormat` was renamed to `RowJdbcOutputFormat`. This change has up till
            # now not been implemented in the python Flink library.
            # https://stackoverflow.com/questions/78960829/java-lang-nosuchmethodexception-in-python-flink-jdbc

            # As you see, java `JdbcSink` connector class has different shape from Python `JdbcSink` connector.
            # In Java code, `jdbcSink` object is generated from `JdbcSinkBuilder` class, but in Python it is not.
            # I think these errors are due to API version mismatch. Any idea to solve these errors?
            # https://stackoverflow.com/questions/79604252/issue-with-pyflink-api-code-for-inserting-data-into-sql

            JdbcConnectionOptions.JdbcConnectionOptionsBuilder()
            .with_url('jdbc:postgresql://localhost:5432/crate')
            .with_driver_name('org.postgresql.Driver')
            .with_user_name("crate")
            .with_password("crate")
            .build(),
            JdbcExecutionOptions.builder()
            .with_batch_interval_ms(1000)
            .with_batch_size(200)
            .with_max_retries(5)
            .build()
        )
    )

    # Execute pipeline.
    env.execute()


if __name__ == '__main__':
    logging.basicConfig(
        format='[%(asctime)s] [%(levelname)s] [%(filename)s:%(lineno)d] - %(message)s',
        level=logging.DEBUG
    )

    logger.info("Start")
    main()
    logger.info("Ready")
